fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedFieldInfosFormat.java

```diff
@@ -20,28 +20,67 @@
 
 package com.apple.foundationdb.record.lucene.codec;
 
+import com.apple.foundationdb.record.lucene.directory.FDBDirectory;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.lucene.codecs.FieldInfosFormat;
 import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexFileNames;
 import org.apache.lucene.index.SegmentInfo;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.store.IOContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.BitSet;
+import java.util.List;
 
 /**
  * This class attempts to reduce the number of .fnm entries when numerous segments are present in FDB.
  *
  */
 public class LuceneOptimizedFieldInfosFormat extends FieldInfosFormat {
+    public static final int DEFAULT_MAXIMUM_FIELD_INFO_CACHE_SIZE = 128;
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneOptimizedFieldInfosFormat.class);
 
     private final FieldInfosFormat fieldInfosFormat;
 
+    private final Cache<Pair<Directory, BitSet>, FieldInfos> fieldInfosCache;
```

Comment:
@ScottDugas  pointed out a potential issue with the combination of RYW cache and the FieldsInfos cache:

Consider the following for a single transaction
-- Lookup FieldInfos 
-- Lookup fails in cache
-- Lookup fails in DB
-- New FieldInfos created, written to DB (is now in FDB RYW cache)
-- Another thread tries to read the FIeldInfos, gets them from RYW cache, populates FieldInfosCache
-- Transaction fails to commit

Since the FieldInfos cache is global (now that the Codec is a singleton), the data is available in the cache even though it never made it to the DB. This can lead to inconsistencies as other instances may have as different representation for the same fielding bites.

A potential solution is to move the cache to the `DirectoryWrapper` and make it transaction scoped.

Reply:
Null Data is not cached thus the asMap()